src/attributes/diagnostics.md

```diff
@@ -83,6 +85,56 @@ pub mod m3 {
 > [command-line][rustc-lint-cli], and also supports [setting
 > caps][rustc-lint-caps] on the lints that are reported.
 
+All lint attributes support an additional `reason` parameter, to give context why
+a certain attribute was added. This reason will be displayed as part of the lint
+message, if the lint is emitted at the defined level.
+
+```rust
+use std::path::PathBuf;
+
+pub fn get_path() -> PathBuf {
+    #[allow(unused_mut, reason = "this is only modified on some platforms")]
+    let mut file_name = PathBuf::from("git");
+
+    #[cfg(target_os = "windows")]
+    file_name.set_extension("exe");
+
+    file_name
+}
+```
+
+### Lint expectations
+
+With the `#[expect]` attributes lints can be expected in a certain scope. If
+this expectation is not fulfilled a new warning is emitted to the user. The
+lint levels can be overridden with other lint attributes as usual.
```

Comment:
This sentence wasn't really clear to me what it meant. I might suggest breaking the example up into two parts. The first example can show where `unfulfilled_lint_expectations` is unfulfilled. Then, after that example, write a small paragraph and a second example that explains how *changing* the level of a lint within an expectation scope will override the `expect` attribute. Perhaps something like:

----

If the level of a lint is modified within an `expect` scope with a `warn` or `deny` attribute, then triggering that lint does *not* satisfy the `expect` attribute, and the `unfulfilled_lint_expectations` will also trigger. Similarly, squelching a lint with `allow` will also prevent the lint from being fulfilled.

----

This behavior seems kinda confusing to me, and it is not really clear to me how it is working, or if that is really intended behavior? 

I might also include some text explaining the interaction with `forbid`, since that also isn't entirely clear to me. I believe it is something like "if a lint has previously been marked with `forbid`, then it is not allowed to use the `expect` attribute with that lint".

Reply:
> This behavior seems kinda confusing to me, and it is not really clear to me how it is working, or if that is really intended behavior?

This is intended behavior, see https://github.com/rust-lang/rust/issues/54503#issuecomment-1179701134

I've added more examples and documentation. :upside_down_face: